src/Storages/ObjectStorage/StorageObjectStorageCluster.cpp

```diff
@@ -268,6 +268,55 @@ void StorageObjectStorageCluster::updateQueryToSendIfNeeded(
     if (cluster_name_in_settings)
     {
         configuration->addStructureAndFormatToArgsIfNeeded(args, structure, configuration->format, context, /*with_structure=*/true);
+
+        /// Convert to old-stype *Cluster table function.
+        /// This allows to use old clickhouse versions in cluster.
+        static std::unordered_map<std::string, std::string> function_to_cluster_function = {
+            {"s3", "s3Cluster"},
+            {"azureBlobStorage", "azureBlobStorageCluster"},
+            {"hdfs", "hdfsCluster"},
+            {"iceberg", "icebergS3Cluster"},
+            {"icebergS3", "icebergS3Cluster"},
+            {"icebergAzure", "icebergAzureCluster"},
+            {"icebergHDFS", "icebergHDFSCluster"},
+            {"deltaLake", "deltaLakeCluster"},
+            {"hudi", "hudiCluster"},
+        };
+
+        ASTFunction * table_function = extractTableFunctionFromSelectQuery(query);
+
+        auto p = function_to_cluster_function.find(table_function->name);
+        if (p == function_to_cluster_function.end())
+        {
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Can't find cluster name for table function {}",
+                table_function->name);
+        }
+
+        table_function->name = p->second;
+
+        auto cluster_name = getClusterName(context);
+        auto cluster_name_arg = std::make_shared<ASTLiteral>(cluster_name);
+        args.insert(args.begin(), cluster_name_arg);
```

Comment:
What if the user does something creepy like:

`select * from s3Cluster("cluster1"...) SETTINGS object_storage_cluster="cluster2";`